src/sequence_batch_scheduler.cc

```diff
@@ -1215,12 +1387,39 @@ DirectSequenceBatch::DirectSequenceBatch(
 
 DirectSequenceBatch::~DirectSequenceBatch()
 {
-  // Signal the scheduler thread to exit...
+  // Wait until all queued requests begin execution.
   {
-    std::unique_lock<std::mutex> lock(mu_);
-    scheduler_thread_exit_ = true;
+    std::unique_lock<std::mutex> lk(mu_);
+    queues_cv_.wait(lk, [this] {
+      for (uint32_t seq_slot = 0; seq_slot < queues_.size(); seq_slot++) {
+        if (!queues_[seq_slot].empty()) {
+          LOG_VERBOSE(1) << "Waiting for slot " << seq_slot
+                         << " to begin execution before exiting";
+          return false;
+        }
+      }
+      return true;
+    });
+  }
+
+  // Wait until the last enqueued payload completes execution.
+  {
+    std::unique_lock<std::mutex> lk(payload_mu_);
+    payload_cv_.wait(lk, [this] {
+      if (!exec_complete_ || curr_payload_->RequestCount() > 0) {
+        LOG_VERBOSE(1) << "Waiting for current payload to complete execution "
+                          "before exiting";
+        return false;
+      }
+      return true;
+    });
   }
 
+  // Signal the scheduler thread to exit.
+  {
+    std::lock_guard<std::mutex> lk(mu_);
+    scheduler_thread_exit_ = true;
```

Comment:
std::atomic<bool> would be simpler and safer here. These seem to be inconsistently protected.

Reply:
Yes, updated on commit: [Use atomic bool for thread exit vars](https://github.com/triton-inference-server/core/pull/207/commits/d875d79519ccdd3d0877161dea58ea4a02b503ea)